keras/backend/mxnet_backend.py

```diff
@@ -3931,18 +3931,50 @@ def __getitem__(self, in_slice):
         # Convert it to a tuple iterator for single dimensional bias Tensors
         if not isinstance(in_slice, (list, tuple)):
             in_slice = (in_slice,)
+
+        # MXNet sym.slice() operator does not support slicing the complete axis with '-1' indexing.
+        # See here for more details - https://github.com/awslabs/keras-apache-mxnet/issues/113
+        # This is a workaround solution, where:
+        # 1. We slice all the elements with indexing -1
+        # 2. Squeeze the axis to get the effect of slicing the complete axis.
+        # Ex: if input.shape (2,2,3) and indexing is like input[:, -1, :]
+        #     1. After step 1, result = (2,1,3)
+        #     2. After step 2, result = (2,3) which is same as Numpy and other backend behavior.
+        slice_axis = False
+        sliced_dim = [d for d in range(len(in_slice)) if in_slice[d] == -1]
         for i in in_slice:
             if isinstance(i, int):
-                begin.append(i)
-                end.append(i + 1)
+                # Want to slice the complete axis
+                if i == -1:
```

Comment:
Testing some corner cases and found out both numpy and keras-tf can handle slicing with negative index until the length of that axis, we should handle all negative index cases.

```
np_data = np.random.randn(5, 5, 5)
var = K.variable(np_data, dtype=np_data.dtype)
>>> np.array_equal(K.eval(var[:, 1, :]), np_data[:,1,:])
True
>>> np.array_equal(K.eval(var[:, -1, :]), np_data[:,-1,:])
True
>>> np.array_equal(K.eval(var[:, -2, :]), np_data[:,-2,:])
True
>>> np.array_equal(K.eval(var[:, -3, :]), np_data[:,-3,:])
True
>>> np.array_equal(K.eval(var[:, -4, :]), np_data[:,-4,:])
True
>>> np.array_equal(K.eval(var[:, -5, :]), np_data[:,-5,:])
True
```

Reply:
Good point. Thanks.
1. This cannot be supported with at Keras-MXNet backend level, because we cannot infer shape and validate if it is a PlaceHolder and slicing is in Batch dimension. 
2. -2, -3 and such indexing is very rare, and it is highly likely -1 is used for slicing the axis. Hence, I would rather leave this as not supported (till mxnet slicing supports it)

Let me know your suggestions.

Reply:
Correct me if I am wrong: with the current code, following is the output of -3 index, you just need to squeeze it. The output is almost correct. 
```
>>> K.eval(var[:, -3, :])
array([[[ 1.05196469, -0.36612039,  0.55118293, -1.24901296,
          0.1445943 ]],

       [[ 0.38754396, -1.38874179, -1.73223019,  2.03238142,
          0.98131609]],

       [[-0.85328576, -0.18601252, -0.43296869,  0.19665155,
          0.11326693]],

       [[-0.64659878,  1.53067163,  0.28361218,  0.23279734,
         -0.198544  ]],

       [[ 0.26599282,  2.07316213, -0.2963127 ,  1.27382705,
         -0.93112912]]])
>>> np_data[:,-3,:]
array([[ 1.05196469, -0.36612039,  0.55118293, -1.24901296,  0.1445943 ],
       [ 0.38754396, -1.38874179, -1.73223019,  2.03238142,  0.98131609],
       [-0.85328576, -0.18601252, -0.43296869,  0.19665155,  0.11326693],
       [-0.64659878,  1.53067163,  0.28361218,  0.23279734, -0.198544  ],
       [ 0.26599282,  2.07316213, -0.2963127 ,  1.27382705, -0.93112912]])
>>> K.eval(var[:, -3, :]).shape
(5, 1, 5)
>>> np_data[:,-3,:].shape
(5, 5)

```

Reply:
You see that because, it is not actually slicing the axis as expected but treating it as begin=(None, -3, None), end = (None, -2, None).
So in effect MXNet is not understanding it needs to slice axis, hence the dimension is 1.